odoo/uninstall.py

```diff
@@ -0,0 +1,1055 @@
+                if not field.relational and not has_inverse:
+                    continue
+
+                if self.reverse:
+                    fields.setdefault(field.comodel_name, set())
+                    fields[field.comodel_name].add(field)
+                else:
+                    fields.setdefault(field.model_name, set())
+                    fields[field.model_name].add(field)
+
+        return fields
+
+
+class Uninstaller:
+    def __init__(self, data, constraints, relations, level=2):
+        self.env = data.env
+        self.pool = data.pool
+        self.graphs = [
+            PrimalGraph(constructor=lambda x: x) for index in range(0, level)
+        ]
+        self.node_lookup = {}
+        self.model_nodes = {}
+        self.extra_steps = {}
+
+        # Do not delete records that have other external ids (and thus do not
+        # belong to the modules being installed).
+        self.env.cr.execute('''
+            SELECT included_data.id
+            FROM ir_model_data AS included_data
+            LEFT JOIN ir_model_data AS excluded_data
+            ON included_data.res_id = excluded_data.res_id
+            AND included_data.model = excluded_data.model
+            AND excluded_data.id NOT IN %s
+            WHERE included_data.id IN %s
+            AND excluded_data.id IS NULL
+        ''', (tuple(data.ids) or (None,), tuple(data.ids) or (None,)))
+        data = data.browse(rec[0] for rec in self.env.cr.fetchall())
+        data = data.sorted(key='model')
+
+        self.data_by_model = {
+            model: self.env[model].browse(ref.res_id for ref in refs)
+            for model, refs in groupby(data, lambda ref: ref.model)
+        }
+
+        self.data_lookup = {
+            self.env[ref.model].browse(ref.res_id): ref.name for ref in data
+        }
+
+        Model = self.env['ir.model']
+        Field = self.env['ir.model.fields']
+        Constraint = self.env['ir.model.constraint']
+        Relation = self.env['ir.model.relation']
+
+        self.models = self.data_by_model.get('ir.model', Model)
+        self.fields = self.data_by_model.get('ir.model.fields', Field)
+
+        self.simple_fields = self.fields.filtered(lambda field: all([
+            field.ttype not in ('one2many', 'many2one', 'many2many'),
+            not field.related_field_id and not field.related,
+        ]))
+
+        # Add missing relation tables. For example, the relation table for an
+        # inherited many2many field defined in another module will not be
+        # marked for removal.
+        relations |= Relation.search([
+            ('name', 'in', self.fields.mapped('relation_table')),
+            ('name', 'not in', Field.search([
+                ('id', 'not in', self.fields.ids),
+                ('relation_table', 'in', self.fields.mapped('relation_table')),
+                ('relation_table', '!=', False),
+            ]).mapped('relation_table')),
+        ])
+
+        if len(constraints) > 0:
```

Comment:
The idiomatic way is
```suggestion
        if constraints:
```